File: exercism/c/complex-numbers/src/complex_numbers.c
#include "complex_numbers.h"
#include <math.h>

complex_t c_add(complex_t a, complex_t b)
{
  complex_t z = {.real = a.real + b.real, .imag = a.imag + b.imag};
  return z;
}

complex_t c_sub(complex_t a, complex_t b)
{
  complex_t z = {.real = a.real - b.real, .imag = a.imag - b.imag};
  return z;
}

complex_t c_mul(complex_t a, complex_t b)
{
  complex_t z = {.real = a.real * b.real - a.imag * b.imag,
                 .imag = a.imag * b.real + a.real * b.imag};
  return z;
}

complex_t c_div(complex_t a, complex_t b)
{
  double denom = b.real * b.real + b.imag * b.imag;
  complex_t z = {.real = (a.real * b.real + a.imag * b.imag) / denom,
                 .imag = (a.imag * b.real - a.real * b.imag) / denom};
  return z;
}

double c_abs(complex_t x) { return sqrt(x.real * x.real + x.imag * x.imag); }

complex_t c_conjugate(complex_t x)
{
  complex_t z = {.real = x.real, .imag = -x.imag};
  return z;
}

double c_real(complex_t x) { return x.real; }

double c_imag(complex_t x) { return x.imag; }

complex_t c_exp(complex_t x)
{
  double e_x = exp(x.real);
  complex_t z = {.real = e_x * cos(x.imag), .imag = e_x * sin(x.imag)};
  return z;
}

File: exercism/python/complex-numbers/complex_numbers.py
import math

class ComplexNumber(object):

    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        return ComplexNumber(
                self.real + other.real,
                self.imaginary + other.imaginary)

    def __mul__(self, other):
        return ComplexNumber(
                self.real * other.real - self.imaginary * other.imaginary,
                self.imaginary * other.real + self.real * other.imaginary)

    def __sub__(self, other):
        return self + ComplexNumber(-1, 0) * other

    def __truediv__(self, other):
        scale = ComplexNumber(1/(other.real**2 + other.imaginary**2), 0)
        return scale * ComplexNumber(
                self.real * other.real + self.imaginary * other.imaginary,
                self.imaginary * other.real - self.real * other.imaginary)

    def __abs__(self):
        return math.sqrt(self.real**2 + self.imaginary**2)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        return ComplexNumber(math.exp(self.real), 0) * ComplexNumber(
                math.cos(self.imaginary),
                math.sin(self.imaginary))
